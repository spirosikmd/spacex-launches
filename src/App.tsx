import React, { Suspense } from 'react';
import PropTypes from 'prop-types';
import { Router } from '@reach/router';
import CssBaseline from '@material-ui/core/CssBaseline';
import {
  withStyles,
  Theme,
  createStyles,
  WithStyles,
} from '@material-ui/core/styles';
import Loader from './Loader';
import TopBar from './TopBar';
import Footer from './Footer';

const HomePage = React.lazy(() => import('./HomePage'));
const NotFoundPage = React.lazy(() => import('./NotFoundPage'));
const LaunchPage = React.lazy(() => import('./LaunchPage'));

const styles = (theme: Theme) =>
  createStyles({
    main: {
      padding: theme.spacing(2),
    },
  });

const App = ({ classes }: WithStyles<typeof styles>) => {
  return (
    <>
      <TopBar />
      <main className={classes.main}>
        <CssBaseline />
        <Suspense fallback={<Loader />}>
          <Router>
            <NotFoundPage default />
            <HomePage path="/" />
            <LaunchPage path="/launches/:flightNumber" />
          </Router>
        </Suspense>
      </main>
      <Footer />
    </>
  );
};

App.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(React.memo(App));
