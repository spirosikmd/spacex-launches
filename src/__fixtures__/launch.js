export function createLaunchResponse(data) {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data) {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
