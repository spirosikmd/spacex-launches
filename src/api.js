function sanitizeLaunchesResponse(launches) {
  return launches.map(launch => ({
    flightNumber: launch.flight_number,
    utcDate: new Date(launch.launch_date_utc),
    isSuccessful: launch.launch_success === true && launch.upcoming === false,
    isFailed: launch.launch_success === false && launch.upcoming === false,
    isUpcoming: launch.upcoming === true,
    isTentative: launch.is_tentative === true,
    details: launch.details,
    missionName: launch.mission_name,
    missionIds: launch.mission_id || [],
  }));
}

export async function getLaunches({ sortOrder, sortField, filter }) {
  const response = await fetch(
    `https://api.spacexdata.com/v3/launches?filter=${filter}&sort=${sortField}&order=${sortOrder}`,
    {
      headers: {
        'Content-Type': 'application/json',
      },
    }
  );
  const launches = await response.json();
  return sanitizeLaunchesResponse(launches);
}
