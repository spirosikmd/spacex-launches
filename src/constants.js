export const ASC = 'asc';
export const DESC = 'desc';
export const FLIGHT_NUMBER_FIELD = 'flight_number';
export const UTC_DATE_FIELD = 'launch_date_utc';
export const FIELDS = `${FLIGHT_NUMBER_FIELD},${UTC_DATE_FIELD},launch_success,upcoming,is_tentative,details,mission_name,mission_id`;
