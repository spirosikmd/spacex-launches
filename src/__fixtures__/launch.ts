import { LaunchResponseData, LaunchData } from '../api';

export function createLaunchResponse(
  data: LaunchResponseData
): LaunchResponseData {
  return {
    flight_number: 1,
    mission_name: 'Test',
    launch_date_utc: '2019-01-01T18:16:00.000Z',
    launch_success: true,
    upcoming: false,
    is_tentative: false,
    details: 'Details',
    mission_id: ['id1'],
    links: {
      mission_patch: 'missionPatch',
    },
    ...data,
  };
}

export function createLaunch(data: LaunchData): LaunchData {
  return {
    flightNumber: 1,
    missionName: 'Test',
    utcDate: new Date('2019-01-01'),
    isSuccessful: true,
    isFailed: false,
    isUpcoming: false,
    isInProgress: false,
    isTentative: false,
    details: 'Details',
    missionIds: ['id1'],
    missionPatch: 'missionPatch',
    ...data,
  };
}
