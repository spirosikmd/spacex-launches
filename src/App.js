import React, { PureComponent, Suspense } from 'react';
import { Router } from '@reach/router';
import CssBaseline from '@material-ui/core/CssBaseline';
import { withStyles } from '@material-ui/core/styles';
import { getLaunches } from './api';
import { processLaunches } from './utils';
import { DESC, UTC_DATE_FIELD, FIELDS } from './constants';
import Loader from './Loader';
import TopBar from './TopBar';
import Footer from './Footer';

const Home = React.lazy(() => import('./HomePage'));
const NotFound = React.lazy(() => import('./NotFound'));
const LaunchPage = React.lazy(() => import('./LaunchPage'));

const styles = theme => ({
  main: {
    padding: theme.spacing.unit * 2,
  },
});

export class App extends PureComponent {
  state = {
    launches: [],
    isLoadingLaunches: true,
    isUpdatingLaunches: false,
    error: null,
    showUpcoming: false,
    showSuccessful: true,
    showFailed: true,
    sortOrder: DESC,
    sortField: UTC_DATE_FIELD,
  };

  async componentDidMount() {
    try {
      const { sortField, sortOrder } = this.state;
      const launches = await getLaunches({
        sortField,
        sortOrder,
        filter: FIELDS,
      });
      this.setState({ launches, isLoadingLaunches: false });
    } catch (error) {
      this.setState({ error, isLoadingLaunches: false });
    }
  }

  handleChange = (value, checked) => {
    this.setState({ [value]: checked });
  };

  handleSortChange = (name, value) => {
    this.setState(
      {
        [name]: value,
        isUpdatingLaunches: true,
      },
      async () => {
        const { sortField, sortOrder } = this.state;
        const launches = await getLaunches({
          sortField,
          sortOrder,
          filter: FIELDS,
        });
        this.setState({ launches, isUpdatingLaunches: false });
      }
    );
  };

  render() {
    const { classes } = this.props;
    const {
      launches,
      isLoadingLaunches,
      isUpdatingLaunches,
      showUpcoming,
      showSuccessful,
      showFailed,
      sortOrder,
      sortField,
    } = this.state;

    const processedLaunches = processLaunches(launches, {
      showUpcoming,
      showSuccessful,
      showFailed,
    });

    return (
      <>
        <TopBar />
        <main className={classes.main}>
          <CssBaseline />
          <Suspense fallback={<Loader />}>
            <Router>
              <NotFound default />
              <Home
                path="/"
                launches={launches}
                processedLaunches={processedLaunches}
                showFailed={showFailed}
                showSuccessful={showSuccessful}
                showUpcoming={showUpcoming}
                sortField={sortField}
                sortOrder={sortOrder}
                isUpdatingLaunches={isUpdatingLaunches}
                isLoadingLaunches={isLoadingLaunches}
                onChange={this.handleChange}
                onSortChange={this.handleSortChange}
              />
              <LaunchPage path="/launches/:flightNumber" />
            </Router>
          </Suspense>
        </main>
        <Footer />
      </>
    );
  }
}

export default withStyles(styles)(App);
