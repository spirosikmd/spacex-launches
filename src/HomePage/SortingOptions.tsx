import React, { SyntheticEvent } from 'react';
import PropTypes from 'prop-types';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import { ASC, DESC, FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from '../constants';

function getSortOrderOptions(sortField: string) {
  switch (sortField) {
    case FLIGHT_NUMBER_FIELD:
      return (
        <>
          <option value={ASC}>smaller to bigger</option>
          <option value={DESC}>bigger to smaller</option>
        </>
      );
    case UTC_DATE_FIELD:
      return (
        <>
          <option value={ASC}>old to new</option>
          <option value={DESC}>new to old</option>
        </>
      );
    default:
      return (
        <>
          <option value={ASC}>ascending</option>
          <option value={DESC}>descending</option>
        </>
      );
  }
}

interface SortingOptionsProps {
  sortField: string;
  sortOrder: string;
  onSortChange: (name: string, value: string) => void;
}

const SortingOptions = ({
  sortField,
  sortOrder,
  onSortChange,
}: SortingOptionsProps) => {
  const handleSortChange = (event: SyntheticEvent) => {
    const target = event.currentTarget as HTMLSelectElement;
    onSortChange(target.name, target.value);
  };

  return (
    <Grid container spacing={8}>
      <Grid item>
        <TextField
          id="sort-field"
          select
          label="Sort field"
          value={sortField}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortField',
          }}
          margin="normal"
          variant="outlined"
        >
          <option value={FLIGHT_NUMBER_FIELD}>flight number</option>
          <option value={UTC_DATE_FIELD}>utc date</option>
        </TextField>
      </Grid>
      <Grid item>
        <TextField
          id="sort-order"
          select
          label="Sort order"
          value={sortOrder}
          onChange={handleSortChange}
          SelectProps={{
            native: true,
            name: 'sortOrder',
          }}
          margin="normal"
          variant="outlined"
        >
          {getSortOrderOptions(sortField)}
        </TextField>
      </Grid>
    </Grid>
  );
};

SortingOptions.propTypes = {
  sortField: PropTypes.string.isRequired,
  sortOrder: PropTypes.string.isRequired,
  onSortChange: PropTypes.func.isRequired,
};

export default React.memo(SortingOptions);
