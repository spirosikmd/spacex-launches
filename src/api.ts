import { FLIGHT_NUMBER_FIELD, UTC_DATE_FIELD } from './constants';

const BASE = 'https://api.spacexdata.com/v3';
export const LAUNCHES_BASE = `${BASE}/launches`;
export const FILTER =
  'flight_number,launch_date_utc,launch_success,upcoming,is_tentative,details,mission_name,mission_id';

export interface LaunchData {
  flightNumber: number;
  missionName: string;
  utcDate: Date;
  isSuccessful: boolean;
  isFailed: boolean;
  isUpcoming: boolean;
  isInProgress: boolean;
  isTentative: boolean;
  details: string;
  missionIds: string[];
  missionPatch: string;
}

export interface LaunchResponseData {
  flight_number: number;
  mission_name: string;
  launch_date_utc: string;
  launch_success: boolean;
  upcoming: boolean;
  is_tentative: boolean;
  details: string;
  mission_id: string[];
  links: {
    mission_patch: string;
  };
}

function sanitizeLaunchResponse(launch: LaunchResponseData) {
  return {
    flightNumber: launch.flight_number,
    utcDate: new Date(launch.launch_date_utc),
    isSuccessful: launch.launch_success === true && launch.upcoming === false,
    isFailed: launch.launch_success === false && launch.upcoming === false,
    isUpcoming: launch.launch_success === null && launch.upcoming === true,
    isInProgress: launch.launch_success === null && launch.upcoming === false,
    isTentative: launch.is_tentative === true,
    details: launch.details,
    missionName: launch.mission_name,
    missionIds: launch.mission_id || [],
    missionPatch: (launch.links && launch.links.mission_patch) || '',
  };
}

function sanitizeLaunchesResponse(launches: LaunchResponseData[]) {
  return launches.map(sanitizeLaunchResponse);
}

function getHeaders() {
  return {
    'Content-Type': 'application/json',
  };
}

export function mapSortField(sortField: string) {
  switch (sortField) {
    case FLIGHT_NUMBER_FIELD:
      return 'flight_number';
    case UTC_DATE_FIELD:
    default:
      return 'launch_date_utc';
  }
}

async function handleResponse(response: Response) {
  if (!response.ok) {
    const { error } = await response.json();
    throw new Error(error);
  }

  return response.json();
}

export async function getLaunches({
  sortOrder,
  sortField,
}: {
  sortOrder: string;
  sortField: string;
}) {
  const response = await fetch(
    `${LAUNCHES_BASE}?filter=${FILTER}&sort=${mapSortField(
      sortField
    )}&order=${sortOrder}`,
    {
      headers: getHeaders(),
    }
  );

  const launches = await handleResponse(response);

  return sanitizeLaunchesResponse(launches);
}

export async function getLaunch({ flightNumber }: { flightNumber: number }) {
  const response = await fetch(`${LAUNCHES_BASE}/${flightNumber}`, {
    headers: getHeaders(),
  });

  const launch = await handleResponse(response);

  return sanitizeLaunchResponse(launch);
}
