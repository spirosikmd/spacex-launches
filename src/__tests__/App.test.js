import { App } from '../App';
import StatusFilter from '../StatusFilter';
import SortingOptions from '../SortingOptions';
import { getLaunches } from '../api';
import { processLaunches } from '../launches';
import { createLaunch } from '../__fixtures__/launch';
import { FLIGHT_NUMBER_FIELD, DESC } from '../constants';

jest.mock('../api', () => ({
  getLaunches: jest.fn(),
}));

jest.mock('../launches', () => ({
  processLaunches: jest.fn(),
}));

describe('App', () => {
  let props;

  beforeEach(() => {
    props = {
      classes: {
        main: 'main',
        updatingLoader: 'updatingLoader',
      },
    };
    const launches = [createLaunch()];
    getLaunches.mockReturnValue(Promise.resolve(launches));
    processLaunches.mockReturnValue(launches);
  });

  it('renders info, filters, sorting options and list of launches', async () => {
    expect(await shallow(App, props)).toMatchSnapshot();
  });

  describe('when any status filter changes', () => {
    it('sets the value and checked in the state', async () => {
      const app = await shallow(App, props);
      const statusFilter = app.find(StatusFilter);
      statusFilter.prop('onChange')('value', 'checked');
      expect(app.state().value).toBe('checked');
    });
  });

  describe('when sorting options change', () => {
    it('gets launches with new sorting options', async () => {
      const app = await shallow(App, props);
      const sortingOptions = app.find(SortingOptions);
      await sortingOptions.prop('onSortChange')(
        'sortField',
        FLIGHT_NUMBER_FIELD
      );
      const state = app.state();
      expect(state.isUpdatingLaunches).toBe(false);
      expect(getLaunches).toBeCalledWith({
        sortField: FLIGHT_NUMBER_FIELD,
        sortOrder: DESC,
      });
    });
  });

  describe('when is loading launches', () => {
    it('renders a loader', async () => {
      const app = await shallow(App, props);
      app.setState({ isLoadingLaunches: true });
      expect(app).toMatchSnapshot();
    });
  });

  describe('when is not loading launches', () => {
    describe('when is updating launches', () => {
      it('renders info, filters, sorting options, and a loader', async () => {
        const app = await shallow(App, props);
        app.setState({ isUpdatingLaunches: true });
        expect(app).toMatchSnapshot();
      });
    });
  });
});
