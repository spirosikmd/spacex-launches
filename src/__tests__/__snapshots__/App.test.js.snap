// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders info, filters, sorting options and list of launches 1`] = `
<main
  className="main"
>
  <WithStyles(CssBaseline) />
  <WithStyles(Grid)
    container={true}
    spacing={16}
  >
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <React.Memo
        launches={
          Array [
            Object {
              "details": "Details",
              "flightNumber": 1,
              "isFailed": false,
              "isSuccessful": true,
              "isTentative": false,
              "isUpcoming": false,
              "missionName": "Test",
              "utcDate": Object {
                "toLocaleString": [MockFunction],
              },
            },
          ]
        }
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <React.Memo
        onChange={[Function]}
        showFailed={true}
        showSuccessful={true}
        showUpcoming={false}
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <React.Memo
        onSortChange={[Function]}
        sortField="launch_date_utc"
        sortOrder="desc"
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <React.Memo
        launches={
          Array [
            Object {
              "details": "Details",
              "flightNumber": 1,
              "isFailed": false,
              "isSuccessful": true,
              "isTentative": false,
              "isUpcoming": false,
              "missionName": "Test",
              "utcDate": Object {
                "toLocaleString": [MockFunction],
              },
            },
          ]
        }
      />
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</main>
`;

exports[`App when is loading launches renders a loader 1`] = `
<main
  className="main"
>
  <WithStyles(CssBaseline) />
  <WithStyles(Component) />
</main>
`;

exports[`App when is not loading launches when is updating launches renders info, filters, sorting options, and a loader 1`] = `
<main
  className="main"
>
  <WithStyles(CssBaseline) />
  <WithStyles(Grid)
    container={true}
    spacing={16}
  >
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <React.Memo
        launches={
          Array [
            Object {
              "details": "Details",
              "flightNumber": 1,
              "isFailed": false,
              "isSuccessful": true,
              "isTentative": false,
              "isUpcoming": false,
              "missionName": "Test",
              "utcDate": Object {
                "toLocaleString": [MockFunction],
              },
            },
          ]
        }
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <React.Memo
        onChange={[Function]}
        showFailed={true}
        showSuccessful={true}
        showUpcoming={false}
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <React.Memo
        onSortChange={[Function]}
        sortField="launch_date_utc"
        sortOrder="desc"
      />
    </WithStyles(Grid)>
    <WithStyles(Grid)
      item={true}
      xs={12}
    >
      <div
        className="updatingLoader"
      >
        <WithStyles(Component) />
      </div>
    </WithStyles(Grid)>
  </WithStyles(Grid)>
</main>
`;
